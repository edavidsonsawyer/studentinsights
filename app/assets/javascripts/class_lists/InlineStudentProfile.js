import PropTypes from 'prop-types';
import React from 'react';
import _ from 'lodash';
import {nextGrade} from '../helpers/gradeText';
import SectionHeading from '../components/SectionHeading';
import GenericLoader from '../components/GenericLoader';
import StudentPhotoCropped from '../components/StudentPhotoCropped';
import FeedView from '../feed/FeedView';
import {highlightStyleForKey, userFacingValueForKey} from './highlights';
import {equityChecks, equityCheckFlags} from './ClassroomStats';


// Inline student profile for classroom list creator, shown as a modal
export default class InlineStudentProfile extends React.Component {
  render() {
    const {student} = this.props;
    return (
      <div className="InlineStudentProfile" style={styles.root}>
        <h4 style={{display: 'flex', justifyContent: 'space-between', alignItems: 'center'}}>
          <div style={{display: 'flex', justifyContent: 'flex-start', alignItems: 'center'}}>
            <StudentPhotoCropped studentId={student.id} />
            <a style={{fontSize: 24, fontWeight: 'bold'}} href={`/students/${student.id}`} target="_blank" rel="noopener noreferrer">{student.first_name} {student.last_name}</a>
          </div>
          <a style={{fontSize: 12}} href={`/students/${student.id}`} target="_blank" rel="noopener noreferrer">Open full profile</a>
        </h4>
        <div style={styles.columns}>
          {this.renderSimpleColumn()}
          <div style={{...styles.column, marginTop: 20}}>
            {this.renderFeed()}
          </div>
        </div>
      </div>
    );
  }

  renderSimpleColumn() {
    const {student} = this.props;

    const flags = equityCheckFlags(nextGrade(student.grade));
    return (
      <div style={{...styles.column, marginTop: 10, fontSize: 12}}>
        <SectionHeading>Equity checks</SectionHeading>
        <div style={{display: 'flex', flexDirection: 'row', marginLeft: 10, marginRight: 10}}>
          {_.compact(equityChecks(flags)).map(category => this.renderHeaderCell(category))}
        </div>
      </div>
    );
  }

  renderHeaderCell(params) {
    const {student} = this.props;
    const {label, columnHighlightKey, title} = params;
    const highlightStyle = highlightStyleForKey(student, columnHighlightKey);
    return (
      <div key={label} style={styles.equityCheck}>
        <div style={styles.equityCheckLabel} title={title}>{label}</div>
        <div style={{...styles.equityCheckValue, ...highlightStyle}}>
          {userFacingValueForKey(student, columnHighlightKey)}
        </div>
      </div>
    );
  }

  renderFeed() {
    const {student, fetchProfile} = this.props;
    return (
      <div style={styles.feed}>
        <SectionHeading>Notes for {student.first_name}</SectionHeading>
        <GenericLoader
          promiseFn={() => fetchProfile(student.id)}
          render={this.renderFeedWithData.bind(this)} />
      </div>
    );
  }

  renderFeedWithData(json) {
    return <FeedView feedCards={json.feed_cards} />;
  }
}
InlineStudentProfile.propTypes = {
  fetchProfile: PropTypes.func.isRequired,
  student: PropTypes.object.isRequired
};

const styles = {
  root: {
    margin: 20,
    fontSize: 14
  },
  columns: {
  },
  column: {
  },
  equityCheck: {
    flex: 1,
    display: 'flex',
    flexDirection: 'column'
  },
  equityCheckLabel: {
    flex: 1,
    padding: 5,
    paddingBottom: 0
  },
  equityCheckValue: {
    minHeight: '4em',
    margin: 5,
    padding: 5,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    textAlign: 'center'
  },
  feed: {
    paddingTop: 20
  }
};