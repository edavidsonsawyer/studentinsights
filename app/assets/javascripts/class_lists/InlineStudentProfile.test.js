import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {withDefaultNowContext} from '../testing/NowContainer';
import PerDistrictContainer from '../components/PerDistrictContainer';
import InlineStudentProfile from './InlineStudentProfile';
import profile_json from './fixtures/profile_json';
import students_for_grade_level_next_year_json from './fixtures/students_for_grade_level_next_year_json';

function testStudent(attrs = {}) {
  return {
    ...students_for_grade_level_next_year_json.students[0],
    ...attrs
  };
}

export function testProps(props) {
  return {
    fetchProfile(studentId) {
      return Promise.resolve(profile_json);
    },
    student: testStudent({grade: 'KF'}),
    gradeLevelNextYear: '1',
    ...props
  };
}

function testEl(props) {
  return withDefaultNowContext(
    <PerDistrictContainer districtKey="somerville">
      <InlineStudentProfile {...props} />
    </PerDistrictContainer>
  );
}

it('renders without crashing', () => {
  const el = document.createElement('div');
  const props = testProps();
  ReactDOM.render(testEl(props), el);
});

it('snapshots for gradeLevelNextYear=1', () => {
  const props = testProps({
    student: testStudent({grade: 'KF'}),
    gradeLevelNextYear: '1'
  });
  const tree = renderer
    .create(testEl(props))
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('snapshots for gradeLevelNextYear=2', () => {
  const props = testProps({
    student: testStudent({grade: '1'}),
    gradeLevelNextYear: '2'
  });
  const tree = renderer
    .create(testEl(props))
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('snapshots for gradeLevelNextYear=5', () => {
  const props = testProps({
    student: testStudent({grade: '4'}),
    gradeLevelNextYear: '5',
  });
  const tree = renderer
    .create(testEl(props))
    .toJSON();
  expect(tree).toMatchSnapshot();
});