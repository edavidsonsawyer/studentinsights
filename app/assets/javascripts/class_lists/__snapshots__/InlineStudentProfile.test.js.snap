// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `
<div
  className="InlineStudentProfile"
  style={
    Object {
      "fontSize": 14,
      "margin": 20,
    }
  }
>
  <h4
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "flex-start",
        }
      }
    >
      <div
        className="StudentPhotoCropped"
        style={
          Object {
            "backgroundColor": "#ddd",
            "backgroundImage": "url(/students/1/photo)",
            "backgroundPosition": "center top",
            "backgroundPositionY": -10,
            "backgroundRepeat": "no-repeat",
            "backgroundSize": "150%",
            "border": "1px solid #ddd",
            "borderRadius": 3,
            "height": 58,
            "marginRight": 10,
            "overflow": "hidden",
            "width": 58,
          }
        }
      />
      <a
        href="/students/1"
        rel="noopener noreferrer"
        style={
          Object {
            "fontSize": 24,
            "fontWeight": "bold",
          }
        }
        target="_blank"
      >
        Garfield
         
        Skywalker
      </a>
    </div>
    <a
      href="/students/1"
      rel="noopener noreferrer"
      style={
        Object {
          "fontSize": 12,
        }
      }
      target="_blank"
    >
      Open full profile
    </a>
  </h4>
  <div
    style={Object {}}
  >
    <div
      style={
        Object {
          "fontSize": 12,
          "marginTop": 10,
        }
      }
    >
      <div
        className="SectionHeading"
        style={
          Object {
            "borderBottom": "1px solid #333",
            "padding": 10,
          }
        }
      >
        <h4
          style={
            Object {
              "color": "black",
              "display": "inline",
            }
          }
        >
          Equity checks
        </h4>
      </div>
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "marginLeft": 10,
            "marginRight": 10,
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
            }
          }
        >
          <div
            style={
              Object {
                "flex": 1,
                "padding": 5,
                "paddingBottom": 0,
              }
            }
            title="Students who have an IEP or 504 plan."
          >
            IEP or 504
          </div>
          <div
            style={
              Object {
                "backgroundColor": "white",
                "margin": 5,
                "padding": 5,
                "textAlign": "center",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
            }
          }
        >
          <div
            style={
              Object {
                "flex": 1,
                "padding": 5,
                "paddingBottom": 0,
              }
            }
            title="Students receiving English Learning Services or who have in the pastÂ (FLEP)."
          >
            Limited or FLEP
          </div>
          <div
            style={
              Object {
                "backgroundColor": "white",
                "margin": 5,
                "padding": 5,
                "textAlign": "center",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
            }
          }
        >
          <div
            style={
              Object {
                "flex": 1,
                "padding": 5,
                "paddingBottom": 0,
              }
            }
            title="Students broken down by whether they identify their gender as male, female or nonbinary."
          >
            Gender identity
          </div>
          <div
            style={
              Object {
                "backgroundColor": "rgba(81, 185, 86, 0.5)",
                "margin": 5,
                "padding": 5,
                "textAlign": "center",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
            }
          }
        >
          <div
            style={
              Object {
                "flex": 1,
                "padding": 5,
                "paddingBottom": 0,
              }
            }
            title="Students whose are enrolled in the free or reduced lunch program."
          >
            Reduced lunch
          </div>
          <div
            style={
              Object {
                "backgroundColor": "white",
                "margin": 5,
                "padding": 5,
                "textAlign": "center",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
            }
          }
        >
          <div
            style={
              Object {
                "flex": 1,
                "padding": 5,
                "paddingBottom": 0,
              }
            }
            title="Students who had three or more discipline incidents of any kind during this past school year.  Discipline incidents vary in severity; click on the student's name to see more in their profile."
          >
            Discipline, 3+
          </div>
          <div
            style={
              Object {
                "backgroundColor": "white",
                "margin": 5,
                "padding": 5,
                "textAlign": "center",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
            }
          }
        >
          <div
            style={
              Object {
                "flex": 1,
                "padding": 5,
                "paddingBottom": 0,
              }
            }
            title="A boxplot showing the range of students' latest STAR Math percentile scores.  The number represents the median score.  When clicking, green represents students above the 70th percentile and red represents students below the 30th percentile."
          >
            STAR Math
          </div>
          <div
            style={
              Object {
                "backgroundColor": "white",
                "margin": 5,
                "padding": 5,
                "textAlign": "center",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
            }
          }
        >
          <div
            style={
              Object {
                "flex": 1,
                "padding": 5,
                "paddingBottom": 0,
              }
            }
            title="A boxplot showing the range of students' latest STAR Reading percentile scores.  The number represents the median score.  When clicking, green represents students above the 70th percentile and red represents students below the 30th percentile."
          >
            STAR Reading
          </div>
          <div
            style={
              Object {
                "backgroundColor": "white",
                "margin": 5,
                "padding": 5,
                "textAlign": "center",
              }
            }
          />
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "marginTop": 20,
        }
      }
    >
      <div
        style={
          Object {
            "paddingTop": 20,
          }
        }
      >
        <div
          className="SectionHeading"
          style={
            Object {
              "borderBottom": "1px solid #333",
              "padding": 10,
            }
          }
        >
          <h4
            style={
              Object {
                "color": "black",
                "display": "inline",
              }
            }
          >
            Notes for 
            Garfield
          </h4>
        </div>
        <div
          style={
            Object {
              "padding": 10,
            }
          }
        >
          Loading...
        </div>
      </div>
    </div>
  </div>
</div>
`;
